libipl/p10/ipl0.C

```diff
@@ -377,19 +385,84 @@ static int ipl_alignment_check(void)
 static int ipl_set_ref_clock(void)
 {
 	struct pdbg_target *proc;
+    struct pdbg_target *clock_target;
+    enum pdbg_target_status status;
 	int rc = 0;
+    uint8_t clock_count = 0;
 
 	if (ipl_type() == IPL_TYPE_MPIPL)
 		return -1;
 
 	ipl_log(IPL_INFO, "Istep: set_ref_clock: started\n");
 
+	ipl_log(IPL_INFO, "Istep: soft reset clock, and verify clock status register\n");
+	pdbg_for_each_class_target("oscrefclk", clock_target) {
+        uint8_t data;
+
+        status = pdbg_target_probe(clock_target);
+        assert(status == PDBG_TARGET_ENABLED);
+
+        data = OSC_RESET_CMD;
+        rc = i2c_write(clock_target, 0, OSC_CTL_OFFSET, OSC_RESET_CMD_LEN, &data);
+        if(rc){
+            ipl_log(IPL_ERROR, "soft reset command is failed for clock %d, with rc = %d\n",
+                                pdbg_target_index(clock_target), rc);
+
+		    ipl_error_callback(IPL_ERR_CLK_FAILED);
+
+            return 1;
+        }
+        else{
+            ipl_log(IPL_INFO, "soft reset command is successfull for clock %d\n",
+                                pdbg_target_index(clock_target));
+        }
+
+        sleep(1);
+        rc = i2c_read(clock_target, 0, OSC_STAT_OFFSET, OSC_STAT_REG_LEN, &data);
+        if(rc){
+            ipl_log(IPL_ERROR, "status register read is failed for clock %d, with rc = %d\n",
+                                pdbg_target_index(clock_target), rc);
+
+		    ipl_error_callback(IPL_ERR_CLK_FAILED);
+            return 1;
+        }
+        else{
+            ipl_log(IPL_INFO, "status register value for clock %d is 0x%2X\n",
+                                pdbg_target_index(clock_target), data);
+            if(data & OSC_STAT_ERR_MASK){
+                ipl_log(IPL_ERROR, "Calibration is failed for clock %d, status=0x%2X",
+                                pdbg_target_index(clock_target), data);
+                ipl_error_callback(IPL_ERR_CLK_FAILED);
+                return 1;
+            }
+        }
+        clock_count++;
+    }
+
+    if(clock_count != 1 && clock_count != 2){
+        ipl_log(IPL_ERROR, "Invalid number (%d) of clock target found\n",
+                                clock_count);
+        // This is a software bug. So will not call error callback
+        return 1;
+    }
+
 	pdbg_for_each_class_target("proc", proc) {
 		fapi2::ReturnCode fapirc;
 
 		if (!ipl_is_functional(proc))
 			continue;
 
+        if(clock_count == 2){
+            fapi2::ATTR_CP_REFCLOCK_SELECT_Type clock_select =
+                        fapi2::ENUM_ATTR_CP_REFCLOCK_SELECT_BOTH_OSC0;
+            if(!pdbg_target_set_attribute(
+                    proc, "ATTR_CP_REFCLOCK_SELECT", 1, 1, &clock_select)){
```

Comment:
I think you can directly set that enum value instead of the variable.

Just curiosity: How did you choose `BOTH_OSC0` instead of `BOTH_OSC1`? What is the difference between those?

Reply:
these types are used in
https://github.com/open-power/ipl/blob/11775c0f84a3e62735503294c1ee7ec61fcd4c9a/libphal/phal_pdbg.C#L47.

in openbmc we are planning to use always OSC_0/OSC_1/BOTH_OSC0, since we cant see history of clock failure becuase clock gard is not supported.